KEP-4764: Kubectl supports WatchList to list resources (#4767)
kubernetes/enhancements: keps/sig-cli/4764-kubectl-watchlist/README.md

for user to refactor just "a few commands" into "a batch of complex code". So we are seeking an optimization for `kubectl`. With this optimization, all users of `kubectl` can get free performance improvement.

### Goals

* Without breaking any user interface, add support to `kubectl` for fetching collections using **watch**.
* Make `kubectl` hit the cache, thus reducing impact on `kube-apiserver`.

### Non-Goals

* Change `kubectl` functionality.

## Proposal

Existing `kubectl` subcommands like `get`, `describe` should be able to use a _watch list_ to fetch collections (list resources), rather than directly list resources page by page.

At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.

<<[UNRESOLVED Will **watchlist** replace **list** completely]>>
**watchlist** was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully

> rely on it without other fallback mechanisms.

Reviewer: kube-apiserver will support LIST requests forever for backward compatibility reasons.

But we expect that most clients (and libraries) will actually evolve towards using watchlist once this is graduated to GA.

Reviewer: Yes - I think the things like `kubectl` should eventually migrate to WatchList

some users with restricted permissions may encounter authorization issues. Specifically, users might need
to be granted additional **watch** permissions for the corresponding resources to ensure this mechanism 
works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
additional permission, the old behavior should still work properly. In the worst case, some extra API 
calls might cause a slight delay.

Regarding security, users who already have **list** permissions inherently have access to all objects of this type.
Therefore, I don't believe this change will introduce any additional security issues.

## Design Details

Since support for watch lists was merged in version 1.27, the API server allows clients to include the 
`sendInitialEvents` query parameter when performing **watch** operations. This parameter instructs the 
kube-apiserver to first send the existing members of the collection as `ADDED` events before pushing subsequent 
objects changed events. A special `BOOKMARK` event with annotation `k8s.io/initial-events-end=true` will be sent 
at the end to indicate the completion of the initial object push.

> Therefore, where kubectl originally performed a **list** operation using `List`, we can implement a new

Reviewer: Instead of managing the appropriate parameters in kubectl, you should rely on internal client-go implementation that can decide itself whether watchlist can be usef for a given request.
And you achieve it effectively by setting client-go feature gate.

Author: I agree with  that we should rely on internal `client-go` implementation to decide.  However, due to current implementation issue, kubectl cannot directly use this feature-gate, which is what this KEP attempts to solve.

as we can see https://github.com/kubernetes/kubernetes/blob/fe3772890f650f9bcf020b43dc5a51fab0fa17f4/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go#L108, kubectl use `Do` and `Get` method to list resource, not `List`. so the auto switch mechanism in `client-go` seems will not work here.

I described the introduction of a new environment variable earlier because I thought kubectl needed an independent switch to control it. If this is not necessary, I completely agree to reuse the environment variable switch in `client-go` to achieve this purpose. I will update KEP soon later.

Author: I just updated the KEP to clarify this part. PTAL

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Kubectl supports WatchList to list resources. This will allow requests from kubectl to hit the WatchCache,
> giving a performance boost and reducing impact on the kube-apiserver.

Reviewer 2: ```suggestion
Update `kubectl` to support fetching collections using watch lists. This will allow the API server to serve
those requests from its watch cache, which boosts performance and cuts the impact of those request
on the API server.
```

## Motivation

Currently `client-go` could enable `WatchList` by env and it works great, but kubectl is still a pain. 
`kubectl` doest not have a cache mechanism like `client-go`, so it will do `list`  whenever it is used. Furthermore,
request with pager params can not hit Cache of `kube-apiserver`, every request will go through into `etcd`,
which could push up the load of `kube-apiserver` and make system unstable.

In practice, our users have been writing script to unrestrainedly call `kubectl`, which could lead serious performance
issue. As cluster admininstor, we have few method to eliminate it. APF maybe is an option, but it's hard to configure
and don't solve the problem from the root. In the other side, if we are standing in the shoes of user, most of those 
scripts are simple Bash or Python which just simply glue a few `kubectl` commands. It is too heavy and "not worth" 
for user to refactor just "a few commands" into "a batch of complex code". So we are seeking an optimization for `kubectl`. With this optimization, all users of kubectl can get free performance improvement.

### Goals

> * Introduce WatchList to `kubectl` without breaking any user interface.

Reviewer 2: ```suggestion
* Without breaking any user interface, add support to `kubectl` for fetching collections using **watch**.
```

* Make `kubectl` hit the cache, thus reducing impact on `kube-apiserver`.

### Non-Goals

* Change kubectl functionality.

## Proposal

> `kubectl` commands like `get`, `describe` now should be able to use `WatchList` to list resources, rather than

Reviewer 2: ```suggestion
Existing `kubectl` subcommands like `get`, `describe` should be able to use a _watch list_ to fetch collections (list resources), rather than
```

directly list resources page by page.

At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.

<<[UNRESOLVED Will `WatchList` replace `List` completely]>>
`WatchList` was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
rely on it without other fallback mechanisms.

In my view, due to the performance improvements, we should use `WatchList` by default instead of `List` in the

> long term. For the evolution strategy, we can follow the approach of `client-go` and address it in a later release.

Reviewer 2: A point to capture: external API servers may serve collections but only support the legacy **list** verb to fetch these.

Author: I think this is roughly consistent with kube-apiserver, and the key point is how external API handle **watch** request and `sendInitialEvents` param. 

If an error is returned, the fallback logic should always work. But if external API treat it as an normal **watch** request and not send `k8s.io/initial-events-end=true` event, thing goes wrong. 

If external API uses `k8s.io/apiserver` lib, it can get **watchlist** support for free by upgrade the lib to the latest.

<<[/UNRESOLVED]>>

### User Stories (Optional)

#### Story 1

> 

Reviewer 2: ```suggestion

Given that there is a large collection in the control plane, a user wants to fetch that collection using `kubectl get`.

```

As a user, I shouldn't need to worry about the underlying details. I should simply update the version of `kubectl`
to gain the performance improvements for free.

#### Story 2

As a cluster administrator, after upgrading the cluster, I should observe a significant reduction in the requests
penetrating into etcd on the monitoring dashboard. Ideally, the majority of the requests should be cached by the 
apiserver, rather than penetrating into etcd.

### Risks and Mitigations

For users, there are no visible changes to the user interface. However, since different APIs are called, 
some users with restricted permissions may encounter authorization issues. Specifically, users might need

> to be granted additional `watch` permissions for the corresponding resources to ensure this mechanism 

Reviewer 2: ```suggestion
to be granted additional **watch** permissions for the corresponding resources to ensure this mechanism
```

works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
additional permissions, the old behavior should still function properly. In the worst case, some extra API 

> calls might cause a slight delay.

Reviewer 2: Will kubectl cache a hint about whether or not a cluster supports **watch** for fetching collections?

Author: I think it is possible to do so by cache it into a global variable. In this way, during the life cycle of the kubectl process, if we find that **watchlist** is not supported, we can directly skip using the **watchlist** mechanism later.

Reviewer 2: Ah, I meant a (hint) cache that could outlast a particular kubectl process.

Author: Yeah, that's a great idea. maybe we can write a cache file to record if current cluster support **watchlist** or not.  this cache file would be valid for a short period of time afterwards. so other kubectl process could read it and fallback to **list** directly without trying **watchlist**.

But my concern is in the implementation. Currently, there are many references to List (https://github.com/kubernetes/kubernetes/blob/fe3772890f650f9bcf020b43dc5a51fab0fa17f4/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go#L64) and they are deeply nested. Most of them do not involve the acquisition of the cache directory. If we want to add it, we need to change a lot of places. This may cause some omissions and lead to some problems. Maybe we don’t have to keep it exactly the same as the previous cache directory, but find a separate directory?

> Regarding security, users who already have `list` permissions inherently have access to all objects of this type. 

Reviewer 2: ```suggestion
Regarding security, users who already have **list** permissions inherently have access to all objects of this type.
```

Therefore, I don't believe this change will introduce any additional security issues.

## Design Details

> Since the WatchList feature was merged in version 1.27, it allows us to include the sendInitialEvents parameter

Reviewer 2: ```suggestion
Since support for watch lists was merged in version 1.27, the API server allows clients to include the `sendInitialEvents` query parameter
```

> when performing watch operations. This parameter instructs the kube-apiserver to first send the existing objects

Reviewer 2: (nit)
```suggestion
when performing watch operations. This parameter instructs the kube-apiserver to first send the existing
members of the collection
```

as ADDED events before pushing subsequent object change events. A special Bookmark event with annotation 
`k8s.io/initial-events-end=true` is sent at the end to indicate the completion of the initial object push.

> Therefore, where we originally performed a List operation, we can implement a new WatchList by enabling it 

Reviewer 2: ```suggestion
Therefore, where kubectl originally performed a **list** operation using `List`, we can implement a new
`WatchList` function that is enabled
```

> through a new environment variable. In this implementation, we will send a Watch request with the following 

Reviewer 2: ```suggestion
through a new environment variable. In this implementation, we will send a **watch** request with the following
```

parameters:
```go
  // rely on the `ConsistentListFromCache` feature. When this parameter is an empty string, 
  // the apiserver will ensure that the resource version is the latest version currently in etcd.
  //
  // Additionally, the `WatchFromStorageWithoutResourceVersion` feature is disabled by default in 1.30.
  // In this case, the Watch request will correctly return results from the cache.
	listOptions.ResourceVersion = ""

  // allow the sending of bookmark events. Only when this parameter is enabled will 
  // the aforementioned Bookmark event, used to mark the end of the event sequence, be sent.
	listOptions.AllowWatchBookmarks = true

  // enable WatchList, ask kube-apiserver send initial event.
	listOptions.SendInitialEvents = ptr.To(true)
	listOptions.ResourceVersionMatch = metav1.ResourceVersionMatchNotOlderThan
```

> When the Watch request is initiated and begins responding normally, we will aggregate the output events. 

Reviewer 2: ```suggestion
When the **watch** request is initiated and begins responding normally, kubectl will aggregate the output events. 
```

A detail to note here is that depending on whether the `ServerPrint` is enabled, the output event types may vary.
They may be output as `metav1.Table` or as the original object types. 

> Upon encountering the `k8s.io/initial-events-end` event, we will proactively close the `watch` and pass the 

Reviewer 2: ```suggestion
Upon encountering the `k8s.io/initial-events-end` event, kubectl will proactively close the **watch** and pass the
```

aggregated results to the `VisitorFunc` in one go. However, if an exception occurs during the above events, 

> we will directly fallback to the original logic.

Reviewer 2: ```suggestion
kubectl will directly fallback to the original logic, making a new **list** request.
```

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

> This feature will not be enabled by default in kubectl until the dependent featuregate in the apiserver 
> are enabled by default.
> 
> Upgrading requires that the Kubelet's environment variable be configured to `true`, and the apiserver must
> enable `WatchList` and `ConsistentListFromCache`, as well as disable `WatchFromStorageWithoutResourceVersion`. 
> 
> Downgrade requires kubectl environment variable be configured to `false`.

Reviewer 2: ```suggestion
Provided they follow the guidance about permitted version skew, cluster administrators and end users can
upgrade via any supported route. The cluster administrator and the person using `kubectl` do not need to
agree or coordinate about when the client- and server-side features are enabled.
```

### Version Skew Strategy

This feature will not be enabled by default in kubectl until the dependent featuregate in the apiserver 
are enabled by default. However, if users actively enable it while using an older version of the apiserver,
the following scenarios may occur:

> #### `WatchList` is not supported or disabled

Reviewer 2: ```suggestion
#### API server does not support initial events for watches
```

> When the API is on an older version or the dependent features have not yet been enabled, the `sendInitialEvent` 

Reviewer 2: AIUI the API verb is **watch** for a watch list fetch; maybe I'm wrong though? Let's make it clear which we expect.
```suggestion
When the API server is on an older version or the dependent features have not yet been enabled,
the `sendInitialEvent` parameter in the **watch**request will not be correctly processed.
An API server prior to version 1.27 will ignore that query parameter and treat the request as a standard
**watch** request, which may lead to unexpected behavior in kubectl unless correctly handled.
In kube-apiserver version 1.27 and later, where the `WatchList` feature gate is disabled, the API server will respond with an error as expected; kubectl can then fallback to the original **list** verb and associated
client side logic.
```

parameter in the `WatchList` request will not be correctly processed. apiserver prior to version 1.27 will ignore 
this parameter and treat it as a standard `watch` request, which may lead to unexpected behavior in kubectl. 
In apiserver version 1.27 and later, if the WatchList feature is disabled, it will respond with an error as 
expected, and then fallback to the original List logic.

> #### `ConsistentListFromCache` or `WatchFromStorageWithoutResourceVersion` is not supported or configured expectedly

Reviewer 2: ```suggestion
#### `ConsistentListFromCache` or `WatchFromStorageWithoutResourceVersion` feature gates not enabled
```

## Production Readiness Review Questionnaire

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.

> -->

Reviewer 2: This'll need filling in at some point, but isn't required for merging as provisional.

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

> - [x] Feature gate (also fill in values in `kep.yaml`)
>   - Feature gate name: KUBECTL_WATCHLIST
>   - Components depending on the feature gate: kubectl, apiserver
> - [ ] Other
>   - Describe the mechanism:
>   - Will enabling / disabling the feature require downtime of the control
>     plane?
>   - Will enabling / disabling the feature require downtime or reprovisioning
>     of a node?

Reviewer 2: :information_source: I think this suggestion is important. We're not using a feature gate in the Kubernetes sense.

```suggestion
- [ ] Feature gate (also fill in values in `kep.yaml`)
- [x] Other
  - When the environment variable `KUBECTL_WATCHLIST` is set to true, kubectl
    uses a **watch** to fetch collections where possible. When the environment variable
    `KUBECTL_WATCHLIST` is set to false, kubectl uses the **list** verb (legacy behavior).
    If `KUBECTL_WATCHLIST` is unset, the kubectl tool uses its compiled-in default.
    
    Once the feature is generally available, `kubectl` will always attempt to use a **watch**
    first.
  - Will enabling / disabling the feature require downtime of the control
    plane?

    No
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node?

    No
```

###### Does enabling the feature change any default behavior?

kubectl will use `WatchList` mechanism to list resources rather than `List`. 

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

The features can be disabled for a single user by setting the `kubectl` environment

> variable associated with the feature to **OFF**.

Author: well, I have previously referenced other KEPs, such as https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md#v129-remotecommand-subprotocol-exec-cp-and-attach

I changed this into **false** now, what do u thank?

high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

- [ ] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

<!--
Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
implementation difficulties, etc.).
-->

### Dependencies

<!--
This section must be completed when targeting beta to a release.
-->

###### Does this feature depend on any specific services running in the cluster?

> <!--
> Think about both cluster-level services (e.g. metrics-server) as well
> as node-level agents (e.g. specific version of CRI). Focus on external or
> optional services that are needed. For example, if this feature depends on
> a cloud provider API, or upon an external software-defined storage or network
> control plane.
> 
> For each of these, fill in the following—thinking about running existing user workloads
> and creating new ones, as well as about cluster-level services (e.g. DNS):
>   - [Dependency name]
>     - Usage description:
>       - Impact of its outage on the feature:
>       - Impact of its degraded performance or high-error rates on the feature:
> -->

Reviewer 2: ```suggestion
No
```

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

> 

Reviewer 2: :thought_balloon: Is there a way for the denial of the **watch** to include a hint that a **list** would also fail;
for example, this hint could be included whenever `sendInitialEvents` is set and the API server is able to dry-run
check the equivalent **list** verb.

Author: I don't quite understand what you mean by hint? 

When **watchlist** failed, we can give a warning message and then fallback to **list**.  And When rolling back to **list**, the behavior of kubectl should be consistent with the existing behavior, that is, the results of **list** are returned. I'm not sure about what this hint used to tell user?

Author: oh I see the point now. you mean when the **watchlist** fails, the error is returned with a hint, to tell the client **list** would also fail. so it's unnecessary for client to try **list** again. Did I understand correctly?

I don't believe that we have a easy solution for this. As in my understand, this case can only happen when user have no permission to **list** resource. however, URL query parameters are decoded in the `Handler`, but authz happens before that, in the `Filters`. 

A simpler and feasible solution is to initiate a `SubjectAccessReview` during **watchlist** and return the result as Hint. However, this will bring additional consumption to all **watchlist**, and it seems to be not much different from the client re-initiating the **list** request. After all, if the failure is caused by authz, the **list** has not actually started processing at this time, and it will not consume too many resources.

Author: or we can consider the **watch** permission as a superset of the **list** permission. After all, users with the **watch** permission actually already have the **list** permission. In this case, as long as the user can pass the **watch** authz, they will be able to perform subsequent **list** operations for sure. 

However, this is an incompatible change. I am not sure it is appropriate to discuss it in this KEP.

Reviewer 2: Kind of. `WatchList` is a client method not a logical API verb, though (so `WatchList` calls either **watch** or **list**).

Reviewer 2: > initiate a SubjectAccessReview during watchlist and return the result as Hint

I was thinking of this, perhaps with a fast timeout (eg end the attempt after 10ms, and in that case don't include a hint)

Author: I agree with that the semantics is clearer indeed, if there is only one API call. And it's make sense to initiate a `SubjectAccessReview` inside the API server for saving the time of an RPC call between api and client. However fallback should be rara by design, but this machanism will still slow down every **watch** request(even with fast timeout). Although the impact could be very little, is it worth it?

In addition, this seems to be more of an enhancement to `WatchList`, rather than specific to kubectl? If we need to add this optimization, then the `reflector` in `client-go` should also have the same mechanism? I'm not sure if this should be put in `KEP-3157`?

Reviewer 2: Let's resolve this. I used the :thought_balloon: emoji in https://github.com/kubernetes/enhancements/pull/4767#discussion_r1693983129 to try to imply that this was an aside, rather than any kind of blocking feedback.

Author: oh, I see. thanks for your explanation. 😊 

what about if API server initiate a `SubjectAccessReview` and returns a hint only if `WatchList` is disabled or **watch** permission deny? this could minimize the extra authz query.

Reviewer 2: Yes, that's how I'd imagined it.

Author: What we discussed earlier have been updated into the KEP. PTAL

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

> Yes, kubectl will use `watch` rather than `list` to fetch resources list.

Reviewer 2: ```suggestion
Yes; kubectl will use **watch** rather than **list** to fetch resources list.
Sometimes kubectl will then issue a **list** (for example, when the API server has denied the
**watch**.
```

###### Will enabling / using this feature result in introducing new API types?

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->
No

###### Will enabling / using this feature result in any new calls to the cloud provider?

<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->
No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

<!--
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
-->
No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

> The `WatchList` mechanism may introduce some minor delays due to the need to wait for the arrival of the 

Reviewer 2: ```suggestion
Fetching collections using **watch** may introduce some minor delays due to the need to wait for the arrival of the
```

> `initial-events-end` event. According to the current implementation of WatchList, this event can be delayed

Reviewer 2: ```suggestion
`initial-events-end` event. According to the current implementation, this event can be delayed
```

by up to 100 milliseconds.

For small datasets, this delay might be more noticeable. However, due to the performance improvements provided

> by the `WatchList` mechanism, the delays caused by the reasons mentioned above may be offset when dealing

Reviewer 2: ```suggestion
by the _watch list_ mechanism, the delays caused by the reasons mentioned above may be offset when dealing
```

- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

N/A

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

1. We can continue to use the old previous behavior
When retrieving large amounts of data based on a list, pagination is necessary. However, during the pagination process,
the view retrieved in subsequent pages must be consistent with the version at the time of the first page. If such 
requests are to hit the cache, the API needs to implement a multi-version caching mechanism internally. 
Currently, only the EventBuffer in the WatchCache within the API offers a similar mechanism. Implementing a 
multi-version cache for List requests would undoubtedly not only require more memory but also make the implementation 

> more complex.

Author: thanks, fixed

## Infrastructure Needed (Optional)

<!--
Use this section if you need things from the project/SIG. Examples include a
new subproject, repos requested, or GitHub details. Listing these here allows a
SIG to get the process for these resources started right away.

> -->

Author: fixed

to be granted additional **watch** permissions for the corresponding resources to ensure this mechanism 
works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
additional permission, the old behavior should still work properly. In the worst case, some extra API 
calls might cause a slight delay.

Regarding security, users who already have **list** permissions inherently have access to all objects of this type.
Therefore, I don't believe this change will introduce any additional security issues.

## Design Details

Since support for watch lists was merged in version 1.27, the API server allows clients to include the 
`sendInitialEvents` query parameter when performing **watch** operations. This parameter instructs the 
kube-apiserver to first send the existing members of the collection as `ADDED` events before pushing subsequent 
objects changed events. A special `BOOKMARK` event with annotation `k8s.io/initial-events-end=true` will be sent 
at the end to indicate the completion of the initial object push.

This mechanism has been integrated into `client-go` in 1.31, and users who use typed or dynamic client can get

> this for free by setting env `KUBE_FEATURE_WatchListClient=true`. But unfortunately, in current implementation, 

Reviewer 3: does `kubectl` have a concept of feature gates ?
would it be more practical to set an env var or a command line option to disable a faulty feature ?

Author: Yes, It has its own env feature-gate mechanism. But the basic idea is the same, relying on an env vairble to decide which code path to run. so in my understand, the key point is what the env name we should use.

Reviewer 3: As I user of `kubectl` how do I disable faulty `KUBE_FEATURE_WatchListClient` ?

Author: In `alpha` stage, this feature is disabled by default. User who want to enable it need to set envirable `KUBE_FEATURE_WatchListClient=true` to enable this feature.

In the future, when it enables by default in `beta` stage, user can disable this feature by set env viarable `KUBE_FEATURE_WatchListClient=false` explicitly.

Reviewer 4: It's not quite that neat. What's the expected behavior if your cluster updates to 1.30 but you (e.g. kubectl are still on 1.29?

> `kubectl` doesn't not use dynamic or typed client to list resource, it use raw `rest` client. So the auto switching

Reviewer 3: could `kubectl` actually use `client-go` instead of the raw rest client ?

Author: I wish it could, but in my understanding, there is no easy way to do so. `kubectl` uses resource names instead of kinds to fetch resources. Even if it's possible, it will still be a big change.

mechanism implemented in `client-go` can not work with `kubectl` for now.

> To address the issue, a relatively simple method is to borrow the mechanism in `cleint-go`. Therefore, where 

Reviewer 5: ```suggestion
To address the issue, a relatively simple method is to borrow the mechanism in `client-go`. Therefore, where 
```

Author: fixed
